In [4]:
import pandas as pd
import numpy as np
import glob

A base de atendimentos dos Procons nos dará uma visão sobre o reflexo do comportamento das empresas em relação a seus clientes. É natural que haja chamados de reclamações ou dúvidas dos consumidores em relação às empresas, mas um volume alto (acima da media, por exemplo) pode indicar um comportamento desviante na postura da empresa frente à sociedade.

In [66]:
# IMPORTAÇÃO DOS ARQUIVOS CSV DOS ATENDIMENTOS DE PROCONS DE 2020

path = r'C:\DATASETS\Atendimentos Procons\1-Arquivos Baixados'
all_files = glob.glob(path + "/*.csv")

li = []

for filename in all_files:
    temp_df = pd.read_csv(filename, index_col=None, header=0, sep=';', encoding='UTF8', dtype="str")
    li.append(temp_df)

# Concatena o resultado no dataframe
atendimentos_df = pd.concat(li, axis=0, ignore_index=True)

Dificuldades de encontrar o encoding correto para acentuação. Testado UTF-8, UTF-16, UTF-8-SIG, ANSI, latin-1 e outros. Só deu certo após abrir os arquivos no Notepad++ e convertê-los para UTF-8.

### Processamento/Tratamento de Dados:

##### Verificando colunas e tipos:

In [67]:
atendimentos_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1976923 entries, 0 to 1976922
Data columns (total 27 columns):
AnoAtendimento              object
TrimestreAtendimento        object
MesAtendimento              object
DataAtendimento             object
CodigoRegiao                object
Regiao                      object
UF                          object
CodigoTipoAtendimento       object
DescricaoTipoAtendimento    object
CodigoAssunto               object
DescricaoAssunto            object
GrupoAssunto                object
CodigoProblema              object
DescricaoProblema           object
GrupoProblema               object
SexoConsumidor              object
FaixaEtariaConsumidor       object
CEPConsumidor               object
TipoFornecedor              object
RazaoSocialSindec           object
NomeFantasiaSindec          object
CNPJ                        object
RadicalCNPJ                 object
RazaoSocialRFB              object
NomeFantasiaRFB             object
CodigoCNAEPr

##### Verificando colunas e valores:

In [68]:
atendimentos_df.describe

<bound method NDFrame.describe of         AnoAtendimento TrimestreAtendimento MesAtendimento  \
0                 2020                    1              1   
1                 2020                    1              1   
2                 2020                    1              1   
3                 2020                    1              1   
4                 2020                    1              1   
...                ...                  ...            ...   
1976918           2020                    3              9   
1976919           2020                    3              9   
1976920           2020                    3              9   
1976921           2020                    3              9   
1976922           2020                    3              9   

                 DataAtendimento CodigoRegiao        Regiao  UF  \
0        2020-01-09 13:33:24.000           02      Nordeste  PB   
1        2020-01-09 13:44:25.000           02      Nordeste  PB   
2        2020-01-09 

##### Identificando percentual de valores ausentes por coluna:

In [69]:
(atendimentos_df.isnull().sum() / atendimentos_df.shape[0] * 100).sort_values(ascending=False)

NomeFantasiaRFB             60.754162
DescricaoCNAEPrincipal      20.041347
RazaoSocialRFB              19.731623
CodigoCNAEPrincipal         19.731623
NomeFantasiaSindec          16.006036
CEPConsumidor               11.973506
RadicalCNPJ                 10.205051
CNPJ                        10.205051
RazaoSocialSindec            0.002529
CodigoRegiao                 0.000000
UF                           0.000000
Regiao                       0.000000
TrimestreAtendimento         0.000000
DataAtendimento              0.000000
MesAtendimento               0.000000
DescricaoTipoAtendimento     0.000000
CodigoTipoAtendimento        0.000000
DescricaoProblema            0.000000
CodigoAssunto                0.000000
DescricaoAssunto             0.000000
GrupoAssunto                 0.000000
CodigoProblema               0.000000
GrupoProblema                0.000000
SexoConsumidor               0.000000
FaixaEtariaConsumidor        0.000000
TipoFornecedor               0.000000
AnoAtendimen

### Limpeza e correção dos dados:

##### Limpeza e correção da coluna CNPJ:

A primeira coisa que nos interessa aqui é a coluna CNPJ (que pode ter CPF ou CNPJ, conforme descrito no dicionário de dados). Vamos então analisar a coluna:

In [70]:
atendimentos_df.CNPJ.str.len().value_counts()

14.0    1586796
13.0     154384
12.0      26315
11.0       3522
3.0        3000
6.0         412
10.0        356
9.0         184
5.0         170
4.0          37
8.0           1
Name: CNPJ, dtype: int64

Eram esperados tamanhos de 14 (cnpj) ou 11 (cpf) dígitos. Vamos ver quem são as linhas com valores diferentes disso:

Analisando primeiro os valores com comprimento de 12 e 13 dígitos:

In [72]:
atendimentos_df[(atendimentos_df.CNPJ.str.len() == 12)|(atendimentos_df.CNPJ.str.len() == 13)]

AnoAtendimento TrimestreAtendimento MesAtendimento  \
1402267           2020                    3              8   
1402268           2020                    3              8   
1402290           2020                    3              8   
1402294           2020                    3              8   
1402299           2020                    3              8   
...                ...                  ...            ...   
1976914           2020                    3              9   
1976916           2020                    3              9   
1976917           2020                    3              9   
1976918           2020                    3              9   
1976922           2020                    3              9   

             DataAtendimento CodigoRegiao        Regiao  UF  \
1402267  2020-08-28 06:23:40           03       Sudeste  SP   
1402268  2020-08-26 06:35:13           03       Sudeste  SP   
1402290  2020-08-26 06:35:15           03       Sudeste  SP   
1402294  2020-08-26 06:35:16           03       Sudeste  SP   
1402299  2020-08-26 06:35:17           03       Sudeste  SP   
...                      ...          ...           ...  ..   
1976914  2020-09-11 10:05:45           05  Centro-oeste  MT   
1976916  2020-09-11 10:37:48           05  Centro-oeste  MT   
1976917  2020-09-11 10:43:49           05  Centro-oeste  MT   
1976918  2020-09-11 10:43:49           05  Centro-oeste  MT   
1976922  2020-09-11 14:15:26           05  Centro-oeste  MT   

        CodigoTipoAtendimento                 DescricaoTipoAtendimento  \
1402267                     3  CIP (Carta de Informações Preliminares)   
1402268                     3  CIP (Carta de Informações Preliminares)   
1402290                     3  CIP (Carta de Informações Preliminares)   
1402294                     3  CIP (Carta de Informações Preliminares)   
1402299                     3  CIP (Carta de Informações Preliminares)   
...                       ...                                      ...   
1976914                     3  CIP (Carta de Informações Preliminares)   
1976916                     3  CIP (Carta de Informações Preliminares)   
1976917                     3  CIP (Carta de Informações Preliminares)   
1976918                     3  CIP (Carta de Informações Preliminares)   
1976922                     3  CIP (Carta de Informações Preliminares)   

        CodigoAssunto  ... CEPConsumidor TipoFornecedor  \
1402267            89  ...       8260030              1   
1402268            63  ...      11680000              1   
1402290           236  ...       1229010              1   
1402294            54  ...       5797300              1   
1402299           276  ...       3905050              1   
...               ...  ...           ...            ...   
1976914            63  ...      78055756              1   
1976916           259  ...      78058000              1   
1976917           276  ...      78055756              1   
1976918           276  ...      78055756              1   
1976922            55  ...      78055080              1   

                                         RazaoSocialSindec  \
1402267       MERCADOLIVRE.COM ATIVIDADES DE INTERNET LTDA   
1402268  BV FINANCEIRA SA CREDITO FINANCIAMENTO E INVES...   
1402290                     CENTRAL DE PRODUCOES GWUP S. A   
1402294                                      BANCO CSF S/A   
1402299                                   DECOLAR.COM LTDA   
...                                                    ...   
1976914               FACTA INTERMEDIACAO DE NEGOCIOS LTDA   
1976916                  SKY SERVICOS DE BANDA LARGA LTDA.   
1976917                TUIUTUR VIAGENS E TURISMO LTDA - ME   
1976918                             GOL LINHAS AEREAS S.A.   
1976922             CALCARD ADMINISTRADORA DE CARTOES LTDA   

                                        NomeFantasiaSindec           CNPJ  \
1402267  MERCADOLIVRE.COM.BR / MERCADOSHOPS.COM.BR / MERCA  3361252000134   
1402268  

Foi possível perceber que os casos de 12d e 13d são porque está faltando um ou dois zeros à esquerda na coluna CNPJ. Vamos completar:

In [77]:
atendimentos_df['CNPJ'] = np.where(np.logical_or(atendimentos_df.CNPJ.str.len() == 12, atendimentos_df.CNPJ.str.len() == 13), atendimentos_df.CNPJ.str.zfill(14), atendimentos_df.CNPJ)

In [90]:
# Conferindo resultado:
atendimentos_df.CNPJ.str.len().value_counts()

14.0    1767495
11.0       3522
3.0        3000
6.0         412
10.0        356
9.0         184
5.0         170
4.0          37
8.0           1
Name: CNPJ, dtype: int64

In [89]:
# Conferindo se os RadicalCNPJ estão 'dentro' do CNPJ:
atendimentos_df[(atendimentos_df.CNPJ.str.len()==14)&(atendimentos_df.CNPJ.str[0:8] == atendimentos_df.RadicalCNPJ.astype(str))]

AnoAtendimento TrimestreAtendimento MesAtendimento  \
0                 2020                    1              1   
1                 2020                    1              1   
2                 2020                    1              1   
3                 2020                    1              1   
4                 2020                    1              1   
...                ...                  ...            ...   
1976918           2020                    3              9   
1976919           2020                    3              9   
1976920           2020                    3              9   
1976921           2020                    3              9   
1976922           2020                    3              9   

                 DataAtendimento CodigoRegiao        Regiao  UF  \
0        2020-01-09 13:33:24.000           02      Nordeste  PB   
1        2020-01-09 13:44:25.000           02      Nordeste  PB   
2        2020-01-09 13:44:25.000           02      Nordeste  PB   
3        2020-01-09 13:26:47.000           02      Nordeste  PB   
4        2020-01-09 13:47:01.000           02      Nordeste  PB   
...                          ...          ...           ...  ..   
1976918      2020-09-11 10:43:49           05  Centro-oeste  MT   
1976919      2020-09-11 10:43:49           05  Centro-oeste  MT   
1976920      2020-09-11 10:43:49           05  Centro-oeste  MT   
1976921      2020-09-11 13:49:38           05  Centro-oeste  MT   
1976922      2020-09-11 14:15:26           05  Centro-oeste  MT   

        CodigoTipoAtendimento                 DescricaoTipoAtendimento  \
0                           9            Abertura Direta de Reclamação   
1                           9            Abertura Direta de Reclamação   
2                           9            Abertura Direta de Reclamação   
3                           9            Abertura Direta de Reclamação   
4                           9            Abertura Direta de Reclamação   
...                       ...                                      ...   
1976918                     3  CIP (Carta de Informações Preliminares)   
1976919                     3  CIP (Carta de Informações Preliminares)   
1976920                     3  CIP (Carta de Informações Preliminares)   
1976921                     3  CIP (Carta de Informações Preliminares)   
1976922                     3  CIP (Carta de Informações Preliminares)   

        CodigoAssunto  ... CEPConsumidor TipoFornecedor  \
0                  54  ...      58083535              1   
1                 101  ...      58040270              1   
2                 101  ...      58040270              1   
3                 185  ...      58027430              1   
4                 187  ...      58079140              1   
...               ...  ...           ...            ...   
1976918           276  ...      78055756              1   
1976919           276  ...      78055756              1   
1976920           276  ...      78055756              1   
1976921           183  ...      78000000              1   
1976922            55  ...      78055080              1   

                                         RazaoSocialSindec  \
0                                               FORTBRASIL   
1                                       LOJAS RIACHUELO SA   
2                                           ASUS DO BRASIL   
3         ENERGISA DA PARAÍBA-DISTRIBUIDORA DE ENERGIA S/A   
4                                                 OI MÓVEL   
...                                                    ...   
1976918                             GOL LINHAS AEREAS S.A.   
1976919     CVC BRASIL OPERADORA E AGENCIA DE VIAGENS S.A.   
1976920                     JCNS VIAGENS E TURISMO LTDA-ME   
1976921  AGUAS CUIABA S.A. CONCES. SERV. PUB. AGUA E ES...   
1976922             CALCARD ADMINISTRADORA DE CARTOES LTDA   

         NomeFantasiaSindec            CNPJ RadicalCNPJ  \
0           FIDC FORTBRASIL  29665263000100    29665263 

Pela análise acima, percebemos que a quantidade de linhas com 'CNPJ[0:8] == RadicalCNPJ' é a mesma do value_counts para 14 dígitos após completarmos com 0 à esquerda. Ou seja, 1.767.495 linhas. Significa que o complemente de 0 foi positivo.

Vamos agora avaliar os casos da coluna CNPJ com comprimento < 11 dígitos:

In [92]:
atendimentos_df[(atendimentos_df.CNPJ.str.len() < 11)].CNPJ.value_counts()

191           2999
208000100       98
776000101       40
2135            20
395510          19
              ... 
7316391464       1
37788            1
407542           1
411221           1
7736613877       1
Name: CNPJ, Length: 600, dtype: int64

O CNPJ 191 é bem conhecido, trata-se do Banco do Brasil: 00000000000191. Vamos fazer a correção:

In [93]:
atendimentos_df.CNPJ = np.where(atendimentos_df.CNPJ == '191', '00000000000191', atendimentos_df.CNPJ)

Vamos avaliar os demais casos:

Avaliando o CNPJ '208000100':

In [94]:
atendimentos_df[atendimentos_df.CNPJ == '208000100']

AnoAtendimento TrimestreAtendimento MesAtendimento  \
1418231           2020                    3              7   
1418260           2020                    3              7   
1418271           2020                    3              7   
1418311           2020                    3              7   
1419255           2020                    3              7   
...                ...                  ...            ...   
1851325           2020                    3              9   
1859724           2020                    3              8   
1862598           2020                    3              9   
1879747           2020                    3              7   
1879783           2020                    3              9   

             DataAtendimento CodigoRegiao        Regiao  UF  \
1418231  2020-07-02 10:43:30           05  Centro-oeste  DF   
1418260  2020-07-03 11:18:04           05  Centro-oeste  DF   
1418271  2020-07-03 15:34:52           05  Centro-oeste  DF   
1418311  2020-07-06 20:02:18           05  Centro-oeste  DF   
1419255  2020-07-10 13:30:56           05  Centro-oeste  DF   
...                      ...          ...           ...  ..   
1851325  2020-09-29 10:49:13           05  Centro-oeste  GO   
1859724  2020-08-11 15:46:52           05  Centro-oeste  GO   
1862598  2020-09-11 14:16:44           05  Centro-oeste  GO   
1879747  2020-07-23 14:43:07           05  Centro-oeste  GO   
1879783  2020-09-14 10:59:00           05  Centro-oeste  GO   

        CodigoTipoAtendimento                 DescricaoTipoAtendimento  \
1418231                     2                         Simples Consulta   
1418260                     2                         Simples Consulta   
1418271                     2                         Simples Consulta   
1418311                     3  CIP (Carta de Informações Preliminares)   
1419255                     3  CIP (Carta de Informações Preliminares)   
...                       ...                                      ...   
1851325                     6                   Atendimento Preliminar   
1859724                     3  CIP (Carta de Informações Preliminares)   
1862598                     3  CIP (Carta de Informações Preliminares)   
1879747                     6                   Atendimento Preliminar   
1879783                     6                   Atendimento Preliminar   

        CodigoAssunto  ... CEPConsumidor TipoFornecedor  \
1418231            63  ...           NaN              1   
1418260            69  ...      71680368              1   
1418271            53  ...           NaN              1   
1418311            74  ...      71615000              1   
1419255            54  ...      72251012              1   
...               ...  ...           ...            ...   
1851325            76  ...      75160000              1   
1859724            53  ...           NaN              1   
1862598            53  ...           NaN              1   
1879747            54  ...           NaN              1   
1879783            62  ...      72900000              1   

                  RazaoSocialSindec NomeFantasiaSindec       CNPJ RadicalCNPJ  \
1418231       BANCO DE BRASILIA S/A                BRB  208000100    00000208   
1418260       BANCO DE BRASILIA S/A                BRB  208000100    00000208   
1418271       BANCO DE BRASILIA S/A                BRB  208000100    00000208   
1418311       BANCO DE BRASILIA S/A                BRB  208000100    00000208   
1419255       BANCO DE BRASILIA S/A                BRB  208000100    00000208   
...                             ...                ...        ...         ...   
1851325   BRB BANCO DE BRASILIA S/A                BRB  208000100    00000208   
1859724   BRB BANCO DE BRASILIA S/A                BRB  208000100    00000208   
1862598   BRB BANCO DE BRASILIA S/A                BRB  208000100    00000208   
1879747  BRB - BANCO DE BRASILIA SA                BRB  208000100    00000208   
1879783  BRB - BA

O caso do CNPJ 208000100 é de se completar com zeros à esquerda. Vamos fazer:

In [95]:
atendimentos_df['CNPJ'] = np.where(atendimentos_df.CNPJ == '208000100', atendimentos_df.CNPJ.str.zfill(14), atendimentos_df.CNPJ)

Avaliando o CNPJ '776000101':

In [96]:
atendimentos_df[atendimentos_df.CNPJ == '776000101']

AnoAtendimento TrimestreAtendimento MesAtendimento  \
1419296           2020                    3              7   
1430720           2020                    3              8   
1438115           2020                    3              8   
1484699           2020                    3              9   
1517443           2020                    3              9   
1519733           2020                    3              8   
1564346           2020                    3              7   
1586002           2020                    3              7   
1587744           2020                    3              8   
1613688           2020                    3              9   
1629709           2020                    3              7   
1629740           2020                    3              7   
1645099           2020                    3              9   
1646599           2020                    3              8   
1649196           2020                    3              9   
1658641           2020                    3              8   
1664819           2020                    3              7   
1674802           2020                    3              9   
1677323           2020                    3              7   
1698556           2020                    3              7   
1708091           2020                    3              8   
1715525           2020                    3              7   
1717775           2020                    3              8   
1721364           2020                    3              8   
1737800           2020                    3              8   
1750052           2020                    3              9   
1750054           2020                    3              9   
1753610           2020                    3              8   
1787488           2020                    3              8   
1814168           2020                    3              8   
1820645           2020                    3              7   
1833217           2020                    3              7   
1842561           2020                    3              8   
1847146           2020                    3              9   
1848715           2020                    3              9   
1853812           2020                    3              7   
1891244           2020                    3              7   
1904512           2020                    3              8   
1923310           2020                    3              8   
1959658           2020                    3              9   

             DataAtendimento CodigoRegiao        Regiao  UF  \
1419296  2020-07-13 12:43:02           05  Centro-oeste  DF   
1430720  2020-08-04 11:49:43           02      Nordeste  PE   
1438115  2020-08-31 06:19:08           03       Sudeste  SP   
1484699  2020-09-04 06:15:13           03       Sudeste  SP   
1517443  2020-09-09 06:51:24           03       Sudeste  SP   
1519733  2020-08-05 19:00:13           02      Nordeste  BA   
1564346  2020-07-09 11:44:42           01         Norte  AM   
1586002  2020-07-14 08:29:10           01         Norte  PA   
1587744  2020-08-10 06:42:23           03       Sudeste  SP   
1613688  2020-09-23 07:44:38           03       Sudeste  SP   
1629709  2020-07-26 06:12:02           03       Sudeste  SP   
1629740  2020-07-26 06:12:10           03       Sudeste  SP   
1645099  2020-09-14 10:04:26           03       Sudeste  SP   
1646599  2020-08-05 13:54:10           05  Centro-oeste  GO   
1649196  2020-09-02 10:43:45           05  Centro-oeste  GO   
1658641  2020-08-03 14:03:14           01         Norte  TO   
1664819  2020-07-01 12:50:08           01         Norte  TO   
1674802  2020-09-29 12:29:59           03       Sudeste  SP   
1677323  2020-07-27 13:21:24           03       Sudeste  MG   
1698556  2020-07-31 06:33:25           03       Sudeste  SP   
1708091  2020-08-03 06:27:30           03       Sudeste  SP   
1715525  2020-07-13 13:56:13           02      Nordeste  MA   
1717775  2020-08

Trata-se da mesma situação de precisar completar com zeros à esquerda. Em vez de fazermos isso para cada caso, vamos fazer de forma mais ampla:

In [136]:
# Preenchendo com zero à esquerda, tomando o cuidado de verificar antes se esse preenchimento tornará o CNPJ correspondente ao RadicalCNPJ:
atendimentos_df.CNPJ = np.where((atendimentos_df.CNPJ.str.len()<11)&(atendimentos_df.CNPJ.str.zfill(14).str[0:8] == atendimentos_df.RadicalCNPJ), atendimentos_df.CNPJ.str.zfill(14), atendimentos_df.CNPJ)

Vamos ver se resolveu para todos os casos de CNPJ < 11d:

In [137]:
atendimentos_df[(atendimentos_df.CNPJ.str.len() < 11)].CNPJ.value_counts()

Series([], Name: CNPJ, dtype: int64)

Conforme esperado. Temos então apenas CNPJ 14 digítos e CPF 11 dígitos:

In [139]:
atendimentos_df.CNPJ.str.len().value_counts()

14.0    1771655
11.0       3522
Name: CNPJ, dtype: int64

Fim da correção dos tamanhos na coluna CNPJ.
Vamos agora avaliar os nulos, que conforme vimos anteriormente são 10.205051% das linhas.

In [140]:
atendimentos_df[atendimentos_df.CNPJ.isnull()]

AnoAtendimento TrimestreAtendimento MesAtendimento  \
8                 2020                    1              1   
9                 2020                    1              1   
21                2020                    1              1   
24                2020                    1              1   
27                2020                    1              1   
...                ...                  ...            ...   
1976674           2020                    3              7   
1976675           2020                    3              7   
1976871           2020                    3              9   
1976877           2020                    3              9   
1976915           2020                    3              9   

                 DataAtendimento CodigoRegiao        Regiao  UF  \
8        2020-01-09 14:21:24.000           02      Nordeste  PB   
9        2020-01-09 14:47:35.000           02      Nordeste  PB   
21       2020-01-09 16:28:56.000           02      Nordeste  PB   
24       2020-01-09 17:18:15.000           02      Nordeste  PB   
27       2020-01-10 09:24:38.000           02      Nordeste  PB   
...                          ...          ...           ...  ..   
1976674      2020-07-09 06:43:06           03       Sudeste  SP   
1976675      2020-07-09 06:43:07           03       Sudeste  SP   
1976871      2020-09-02 15:18:32           05  Centro-oeste  MT   
1976877      2020-09-03 11:05:17           05  Centro-oeste  MT   
1976915      2020-09-11 10:05:45           05  Centro-oeste  MT   

        CodigoTipoAtendimento                 DescricaoTipoAtendimento  \
8                           9            Abertura Direta de Reclamação   
9                           9            Abertura Direta de Reclamação   
21                          9            Abertura Direta de Reclamação   
24                          2                         Simples Consulta   
27                          9            Abertura Direta de Reclamação   
...                       ...                                      ...   
1976674                     3  CIP (Carta de Informações Preliminares)   
1976675                     3  CIP (Carta de Informações Preliminares)   
1976871                     3  CIP (Carta de Informações Preliminares)   
1976877                     3  CIP (Carta de Informações Preliminares)   
1976915                     3  CIP (Carta de Informações Preliminares)   

        CodigoAssunto  ... CEPConsumidor TipoFornecedor  \
8                 274  ...           NaN              1   
9                 101  ...      58015400              1   
21                 54  ...      58038451              1   
24                188  ...           NaN              1   
27                101  ...      58068453              1   
...               ...  ...           ...            ...   
1976674           185  ...       6730000              1   
1976675           185  ...       5894380              1   
1976871            54  ...      78075800              1   
1976877           101  ...      78005300              1   
1976915            63  ...      78055756              1   

                                         RazaoSocialSindec  \
8                                      UNIVERSO ONLINE S/A   
9                                                  MISTORE   
21                                    CREDICARD MASTERCARD   
24                                                MAX LINK   
27                                               SMART MIX   
...                                                    ...   
1976674  ELETROPAULO METROPOLITANA ELETRICIDADE DE SAO ...   
1976675  ELETROPAULO METROPOLITANA ELETRICIDADE DE SAO ...   
1976871                                ATACADÃO DOS OCULOS   
1976877                                          PRIME TEC   
1976915                    FACTA EMPRÉSTIMO RÁPIDO E FÁCIL   

                      NomeFantasiaSindec CNPJ RadicalCNPJ RazaoSocialRFB  \
8                                    UO

Percebemos que outras colunas auxiliares, que poderiam nos ajudar a encontrar o CNPJ, também são nulas. Conforme análise no início do notebook, possuem índice de nulos relevante: NomeFantasiaRFB: 60.754162%, RazaoSocialRFB: 19.731623%, NomeFantasiaSindec: 16.006036%.
Por questão de tempo, vamos optar por remover do dataframe as linhas que possuem CNPJ nulo:

In [142]:
atendimentos_df = atendimentos_df.dropna(subset=['CNPJ']).reset_index(drop=True)

Verificando resultados:

In [143]:
print(atendimentos_df.shape)
print(atendimentos_df.CNPJ.str.len().value_counts())

(1775177, 27)
14    1771655
11       3522
Name: CNPJ, dtype: int64


In [144]:
# Definindo formato:
atendimentos_df.CNPJ = atendimentos_df.CNPJ.astype('str')

Fim da limpeza e correção da coluna CNPJ.

Como o que nos interessa nessa base é a quantidade de atendimentos por CNPJ / CPF, não precisamos limpar ou corrigir as demais colunas.

### Consolidação dos atendimentos por CNPJ / CPF:

In [159]:
atendimentos_consolidado_df = atendimentos_df.groupby(['CNPJ']).size().reset_index(name='Atendimentos_Procon')
atendimentos_consolidado_df.describe

<bound method NDFrame.describe of                   CNPJ  Atendimentos_Procon
0       00000000000191                10846
1       00000000000272                    2
2       00000000000515                    1
3       00000000000787                    1
4       00000000001244                    1
...                ...                  ...
130791     99893525187                    2
130792     99925001153                    1
130793     99928159653                    1
130794     99932075191                    1
130795  99999999000191                    3

[130796 rows x 2 columns]>

In [155]:
atendimentos_consolidado_df.sort_values('Atendimentos_Procon', ascending=False)

CNPJ  Atendimentos_Procon
9515    02558157000162                78023
113908  40432544000147                75455
121832  61695227000193                48058
102270  33041260065290                41789
8970    02421421000111                37583
...                ...                  ...
54374   14853656000161                    1
54372   14853312000152                    1
54367   14849345000129                    1
54366   14849292000146                    1
130794     99932075191                    1

[130795 rows x 2 columns]

In [161]:
atendimentos_df[atendimentos_df.CNPJ=='99999999000191']

AnoAtendimento TrimestreAtendimento MesAtendimento  \
23762           2020                    1              1   
51115           2020                    1              3   
59508           2020                    1              3   

               DataAtendimento CodigoRegiao   Regiao  UF  \
23762  2020-01-09 05:45:03.000           03  Sudeste  SP   
51115  2020-03-04 06:11:39.000           03  Sudeste  SP   
59508  2020-03-04 06:20:59.000           03  Sudeste  SP   

      CodigoTipoAtendimento                 DescricaoTipoAtendimento  \
23762                     3  CIP (Carta de Informações Preliminares)   
51115                     3  CIP (Carta de Informações Preliminares)   
59508                     3  CIP (Carta de Informações Preliminares)   

      CodigoAssunto  ... CEPConsumidor TipoFornecedor  \
23762           194  ...       8160520              1   
51115           101  ...      15044160              1   
59508           194  ...       8160520              1   

                          RazaoSocialSindec  \
23762  NÃO EMITIR CIPS OU ABRIR RECLAMAÇÕES   
51115  NÃO EMITIR CIPS OU ABRIR RECLAMAÇÕES   
59508  NÃO EMITIR CIPS OU ABRIR RECLAMAÇÕES   

                             NomeFantasiaSindec            CNPJ RadicalCNPJ  \
23762  LISTA EVITE SITES/EMPRESA NÃO LOCALIZADA  99999999000191    99999999   
51115  LISTA EVITE SITES/EMPRESA NÃO LOCALIZADA  99999999000191    99999999   
59508  LISTA EVITE SITES/EMPRESA NÃO LOCALIZADA  99999999000191    99999999   

      RazaoSocialRFB NomeFantasiaRFB CodigoCNAEPrincipal  \
23762            NaN             NaN                 NaN   
51115            NaN             NaN                 NaN   
59508            NaN             NaN                 NaN   

      DescricaoCNAEPrincipal  
23762                    NaN  
51115                    NaN  
59508                    NaN  

[3 rows x 27 columns]

Foi detectado um CNPJ inválido, vamos removê-lo:

In [162]:
atendimentos_consolidado_df.drop(index=130795, inplace=True)

In [168]:
# Como a coluna CNPJ tem tanto CPF quanto CNPJ, vamos mudar o nome dela:
atendimentos_consolidado_df = atendimentos_consolidado_df.rename(columns={"CNPJ": "CPF_CNPJ"})

Ao final, temos então 130.795 empresas/pessoas físicas que constaram em algum tipo de atendimento nos PROCONs em 2020, sendo 128.352 pessoas jurídicas e 2.443 pessoas físicas, conforme mostrado abaixo:

In [172]:
atendimentos_consolidado_df.CPF_CNPJ.str.len().value_counts()

14    128352
11      2443
Name: CPF_CNPJ, dtype: int64

Vamos salvar o resultado:

In [169]:
atendimentos_consolidado_df.to_csv(r'C:\DATASETS\Atendimentos Procons\atentimentos_procon.csv', index = False)